docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,364 @@
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+**Table of Contents**
+
+- [Overview](#overview)
+- [Cascading deletion with Garbage Collector](#cascading-deletion-with-garbage-collector)
+- [Orphaning the descendants with "orphan" finalizer](#orphaning-the-descendants-with-orphan-finalizer)
+  - [Part I. The finalizer framework](#part-i-the-finalizer-framework)
+  - [Part II. The "orphan" finalizer](#part-ii-the-orphan-finalizer)
+- [Related issues](#related-issues)
+  - [Orphan adoption](#orphan-adoption)
+  - [Upgrading a cluster to support cascading deletion](#upgrading-a-cluster-to-support-cascading-deletion)
+- [End-to-End Examples](#end-to-end-examples)
+  - [Life of a Deployment and its descendants](#life-of-a-deployment-and-its-descendants)
+  - [Life of a ConfigMap](#life-of-a-configmap)
+- [Open Questions](#open-questions)
+- [Considered and Rejected Designs](#considered-and-rejected-designs)
+- [1. Tombstone + GC](#1-tombstone--gc)
+- [2. Recovering from abnormal cascading deletion](#2-recovering-from-abnormal-cascading-deletion)
+
+
+# Overview
+
+Currently most cascading deletion logic is implemented at client-side. For example, when deleting a replica set, kubectl uses a reaper to delete the created pods and then delete the replica set. We plan to move the cascading deletion to the server to simplify the client-side logic. In this proposal, we present the garbage collector which implements cascading deletion for all API resources in a generic way; we also present the finalizer framework, particularly the "orphan" finalizer, to enable flexible alternation between cascading deletion and orphaning.
+
+Goals of the design include:
+* Supporting cascading deletion at the server-side.
+* Centralizing the cascading deletion logic, rather than spreading in controllers.
+* Allowing optionally orphan the dependent objects.
+
+Non-goals include:
+* Releasing the name of an object immediately, so it can be reused ASAP.
+
+# Cascading deletion with Garbage Collector
+
+## API Changes
+
+```
+type ObjectMeta struct {
+	...	
+	ParentReferences []ObjectReference
+}
+```
+
+**ObjectMeta.ParentReferences**: links the resource to the parent resources. For example, a replica set `R` created by a deployment `D` should have an entry in ObjectMeta.ParentReferences pointing to `D`. The link should be set when the child object is created. It can be updated after the creation. For safety reasons, we can add validation rules to restrict what resources could be set as parents. For example, Events will likely be banned from being parents.
```

Comment:
Can you clarify if this only links to direct parents? I mean, does a pod `P` created by `R` link to `D`?